### PR TITLE
Major change: new Address format

### PR DESCRIPTION
**Pull request by vzhovnitsky**

### CODE REVIEW

File: app/components/address/WalletAddress.tsx [hunk 1]
```diff
@@ -28,14 +28,22 @@ export const WalletAddress = memo((props: {
...
+    bounceable?: boolean,
 }) => {
     const toaster = useToaster();
     const network = useNetwork();
     const theme = useTheme();
     const navigation = useTypedNavigation();
     const addToDenyList = useAddToDenyList();
-    const friendlyAddress = props.address.toString({ testOnly: network.isTestnet });
+    const [bounceableFormat,] = useBounceableWalletFormat();
+    const bounceable = useMemo(() => {
```
Comment (dvlkv) [inline app/components/address/WalletAddress.tsx:40]: useMemo for single if?

File: app/fragments/wallet/views/PendingTransactions.tsx [hunk 2]
```diff
@@ -249,6 +255,10 @@ export const PendingTransactions = memo(({ address, viewType = 'main' }: { addre
         });
     }, [setPending]);
 
+    if (pending.length <= 0) {
```
Comment (dvlkv) [inline app/fragments/wallet/views/PendingTransactions.tsx:258]: === 0?

File: app/fragments/NewAddressFormatFragment.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    const { oldAddressString, newAddressString } = useMemo(() => {
+        return {
+            oldAddressString: selectedAccount?.address.toString({ testOnly: network.isTestnet }),
+            newAddressString: selectedAccount?.address.toString({ testOnly: network.isTestnet, bounceable: false })
+        }
+    }, [selectedAccount, network.isTestnet]);
+
+    const switchFormat = useCallback(async () => {
+        // Prefetch contacts contract info for current account at least
+        Object.keys(contacts).forEach((contactEntrie) => {
```
Comment (dvlkv) [inline app/fragments/NewAddressFormatFragment.tsx:36]: contractEntry

File: app/fragments/ContactFragment.tsx [hunk 4]
```diff
@@ -49,6 +50,19 @@ export const ContactFragment = fragment(() => {
     const removeContact = useRemoveContact();
     const safeArea = useSafeAreaInsets();
     const contact = useContact(params.address);
+    const contractInfo = useContractInfo(params.address ?? '');
+    const [bounceableFormat,] = useBounceableWalletFormat();
+
+    const addressString = useMemo(() => {
```
Comment (dvlkv) [inline app/fragments/ContactFragment.tsx:56]: It is not complete address, address short string maybe?

File: app/engine/hooks/transactions/useRawAccountTransactions.ts [hunk 5]
```diff
@@ -231,15 +249,13 @@ export function useRawAccountTransactions(client: TonClient4, account: string, r
                 hash = accountLite.account.last.hash;
             }
 
-            log(`[txns-query] fetching ${lt}_${hash} ${sliceFirst ? 'sliceFirst' : ''}`);
```
Comment (dvlkv) [inline app/engine/hooks/transactions/useRawAccountTransactions.ts:234]: why did you remove this log?

File: app/engine/hooks/transactions/useRawAccountTransactions.ts [hunk 6]
```diff
@@ -196,8 +197,24 @@ function rawTransactionToStoredTransaction(tx: Transaction, hash: string, own: A
...
+        storage.set(key, currentTransactionsVersion);
+        queryClient.removeQueries(Queries.Transactions(account));
+    }
+}
+
+export function useRawAccountTransactions(account: string, refetchOnMount: boolean = false) {
     const { isTestnet } = useNetwork();
+    const client = useClient4(isTestnet);
+
+    invalidateTransactions(account);
```
Comment (dvlkv) [inline app/engine/hooks/transactions/useRawAccountTransactions.ts:217]: invalidateTransactions on every render?

File: app/engine/hooks/transactions/usePendingWatcher.ts [hunk 7]
```diff
@@ -14,13 +14,15 @@ export function usePendingWatcher() {
 
     const v4 = useWalletV4(client, account?.addressString || '');
     const lite = useAccountLite(account?.address || null);
-    const firstTransaction = useRawAccountTransactions(client, account?.addressString || '').data?.pages[0]?.[0];
+    const firstTransaction = useRawAccountTransactions(account?.addressString || '', true).data?.pages[0]?.[0];
 
-    useEffect(() => {
-        const transactionsInSync = firstTransaction?.hash === lite?.last?.hash && (v4.data?.last || 0) >= (lite?.block || 0);
+    const txsInSync = useMemo(() => {
```
Comment (dvlkv) [inline app/engine/hooks/transactions/usePendingWatcher.ts:19]: why to use memo?

File: app/engine/hooks/transactions/useAccountTransactions.ts [hunk 8]
```diff
@@ -36,13 +34,13 @@ export function parseStoredMetadata(metadata: StoredContractMetadata): ContractM
 }
 
 
-export function useAccountTransactions(client: TonClient4, account: string, refetchOnMount: boolean = false): {
+export function useAccountTransactions(account: string, refetchOnMount: boolean = false): {
```
Comment (dvlkv) [inline app/engine/hooks/transactions/useAccountTransactions.ts:37]: let's do this in object for better readability, like: { refetchOnMount: true }

File: app/engine/accountWatcher.ts [hunk 9]
```diff
@@ -33,14 +43,26 @@ export function useBlocksWatcher() {
             lastBlockResolve?.(data.seqno);
             lastBlockResolve = undefined;
             lastBlockPromise = Promise.resolve(data.seqno);
+
+            const lastBlockKey = `lastBlock-${selectedAccount?.addressString || ''}`;
+            const lastBlock = storage.getNumber(lastBlockKey) || data.seqno;
+
+            storage.set(lastBlockKey, data.seqno);
+
+            const accsToSync = [
```
Comment (dvlkv) [inline app/engine/accountWatcher.ts:52]: I do not like this. Let's rollback and test solution with single invalidate call.

**Review by dvlkv — CHANGES_REQUESTED** [review]

File: app/fragments/NewAddressFormatFragment.tsx [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+            if (!queryClient.getQueryData(Queries.ContractInfo(contactEntrie[0]))) {
+                queryClient.prefetchQuery(Queries.ContractInfo(contactEntrie[0]));
```
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/fragments/NewAddressFormatFragment.tsx:38]: We prefetch so that when user goes to contacts tab contacts are displayed in new (or old, depending on selected option) correctly (instead of re-rendering individually when ContactsFragment gets presented)

File: app/fragments/NewAddressFormatFragment.tsx [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            }
+        });
+
+        // Switch the format
+        setBounceableFormat(!bounceableFormat);
```
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/fragments/NewAddressFormatFragment.tsx:43]: ✅

File: app/fragments/NewAddressFormatFragment.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/fragments/NewAddressFormatFragment.tsx]: ✅

File: app/fragments/ContactFragment.tsx (hunk 4 above)
Comment (vzhovnitsky) [reply app/fragments/ContactFragment.tsx]: ✅

File: app/engine/hooks/transactions/useRawAccountTransactions.ts (hunk 5 above)
Comment (vzhovnitsky) [reply app/engine/hooks/transactions/useRawAccountTransactions.ts:234]: ✅

File: app/fragments/NewAddressFormatFragment.tsx (hunk 3 above)
Comment (vzhovnitsky) [reply app/fragments/NewAddressFormatFragment.tsx]: ✅

File: app/engine/hooks/transactions/usePendingWatcher.ts (hunk 7 above)
Comment (vzhovnitsky) [reply app/engine/hooks/transactions/usePendingWatcher.ts]: ✅

File: app/engine/hooks/transactions/useRawAccountTransactions.ts (hunk 6 above)
Comment (vzhovnitsky) [reply app/engine/hooks/transactions/useRawAccountTransactions.ts]: ✅

File: app/engine/hooks/transactions/useAccountTransactions.ts (hunk 8 above)
Comment (vzhovnitsky) [reply app/engine/hooks/transactions/useAccountTransactions.ts]: Sure, you are right ✅

File: app/engine/accountWatcher.ts (hunk 9 above)
Comment (vzhovnitsky) [reply app/engine/accountWatcher.ts]: added new version as we discussed ✅

File: app/engine/hooks/transactions/useRawAccountTransactions.ts [hunk 12]
```diff
@@ -196,12 +199,30 @@ function rawTransactionToStoredTransaction(tx: Transaction, hash: string, own: A
 
 const TRANSACTIONS_LENGTH = 16;
 
-export function useRawAccountTransactions(client: TonClient4, account: string, refetchOnMount: boolean = false) {
+const currentTransactionsVersion = 1;
+const transactionsKey = (account: string) => `transactions-version-${account}`;
```
[the remark this answers is not in the document]
Comment (vzhovnitsky) [reply app/engine/hooks/transactions/useRawAccountTransactions.ts]: ✅

**Review by dvlkv — APPROVED** [review]
